### PR TITLE
Support hhvm 4.73 and xhp 4

### PR DESCRIPTION
**Pull request by lexidor**
This PR makes the required changes to pass the typechecker with hhvm 4.73.
I have not updated the outdated React.createElement() javascript.
I have not formatted the code with a modern hackfmt, to keep the diff clear.
We can do a formatting PR later.

refs #6

### CODE REVIEW

**Comment by fredemmott** [comment]
Sorry for not seeing this sooner; this is a great starting point, but for now I'd rather keep it complete than remove the react support, which is the main purpose for this project. @jjergus, do you agree?

**Comment by jjergus** [comment]
I was only vaguely aware of the existence of this project so don't have much context, but it seems to me that supporting non-React JS projects is still valuable?

Obviously it would be best to support both React and recent HHVM versions (xhp-lib 4), but if I only had to pick one I'd probably say support for recent HHVM/xhp-lib 4 is more important (but again, I don't really have much context).

**Comment by lexidor** [comment]
React support was not removed. I merely marked it as deprecated to inform everyone that the current implementation does not work. I created opened #8.

File: src/x_js-scope.php [hunk 1]
```diff
@@ -8,28 +8,33 @@
...
 
-    $child_waithandles = Vector { };
+    $child_waithandles = vec[];
     foreach ($this->getChildren() as $child) {
-      if ($child instanceof :x:composable-element) {
+      if ($child is x\node) {
         $child->__transferContext($this->getAllContexts());
+        // Can't use Vec\map_async, since we want these Awaitables to start
+        // when the entire tree is x\node only and all contexts have been transferred.
         $child_waithandles[] = (async () ==> await $child->__flushSubtree())();
```
Comment (jjergus) [inline src/x_js-scope.php]: I don't have the context (sorry about the pun :p) to understand what exactly needs to happen here but either the comment is inaccurate or the code is incorrect -- calling the async lambda may (and almost certainly will) eagerly execute some part of it, possibly all of it. Unless I'm missing something, this seems equivalent to map_async.

**Review by jjergus — APPROVED** [review]
Looks good to me except for this one thing

File: src/x_js-scope.php (hunk 1 above)
Comment (lexidor) [reply src/x_js-scope.php]: Yeah, you're right.

`$children = await Vec\map_async($this->getChildren(), $c ==> $c as x\node->__flushSubtree());` would be identical. A non-awaited awaitable is not necessarily not started. So this has no effect on ordering.

Comment (lexidor) [reply src/x_js-scope.php]: Sorry for addressing this 3 months late :shy_laughter:.

**Comment by jjergus** [comment]
Thanks!

**Pull request merged**